go/pkg/cas/upload.go

```diff
@@ -266,25 +332,87 @@ type uploader struct {
 
 // startProcessing adds the item to the appropriate stage depending on its type.
 func (u *uploader) startProcessing(ctx context.Context, ps *PathSpec) error {
-	if !filepath.IsAbs(ps.Path) {
-		return errors.Errorf("%q is not absolute", ps.Path)
+	origPsPtr := ps
+
+	// Do not use os.Stat() here. We want to know if it is a symlink.
+	// TODO(nodir): cache this syscall, perhaps using filemetadat package.
+	info, err := os.Lstat(ps.Path)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	ps = ps.deepCopy()
+	if err := ps.validateAndNormalize(info.Mode()); err != nil {
+		return errors.WithStack(err)
 	}
-	cpy := *ps
-	ps = &cpy
-	ps.Path = filepath.Clean(ps.Path)
 
 	// Schedule a file system walk.
 	u.wgFS.Add(1)
 	u.eg.Go(func() error {
 		defer u.wgFS.Done()
-		// Do not use os.Stat() here. We want to know if it is a symlink.
-		info, err := os.Lstat(ps.Path)
-		if err != nil {
-			return errors.WithStack(err)
+
+		if len(ps.Allowlist) == 0 {
```

Comment:
I wonder if instead you could tweak this case to be treated as if an empty allowlist is treated as a `[]Allowlist{"."}`? It could save quite a few lines of code

Reply:
This is exactly what I did in an earlier version :), but then problem was that then it would mean I'd have to have the "if root, then persist with ps pointer cache key" logic in two places: 
1) when iterating over return value of `missingDirs()` like now
2) but also, in the ps.Allowlist loop.

however, thinking about this a bit more, we have it in two places anyway. So I am happy to change this if you prefer.

Reply:
I _think_ the code readability is worth it the small overhead of creating a tree/waiting on a lock. I _think_ that we wouldn't need to check it in the Allowlist loop, since it'd automatically satisfy the `if dir.relPath == "."` condition on the `missingAncestors` function. Unless I'm missing some other condition (and you concur), I'd prefer the general treatment.

Reply:
done. The UploadInput refactoring incidentally made this much simpler.